### PR TITLE
"Optimizing Stock Price Prediction with LightGBM and CatBoost: Choosing the Right Model for Efficiency and Accuracy"

### PR DESCRIPTION
**Pull request by praveenarjun**
For stock price prediction, you might choose LightGBM or CatBoost depending on the nature of your dataset and specific requirements. If your dataset includes many categorical features, CatBoost might be more suitable. If you need to handle a very large dataset efficiently, LightGBM could be the better choice.

# The main Reason using This model LightGBM or CatBoost because   It has built-in support for categorical features, which can improve performance and accuracy. 

# And These model will reduce time complexity

### CODE REVIEW

**Comment by praveenarjun** [comment]
Hi,

I have reviewed the changes in this pull request. If any changes need to be done or if there are any issues, please make a comment so that it will be beneficial.

Additionally, could you please add the labels "gssoc", "level" and "hacktober"

Thank you!

**Comment by rohitinu6** [comment]
@praveenarjun 
Thank you for your valuable contribution.
Please ensure that you star this repository, your PR will be soon reviewed by the mentors.

**Comment by praveenarjun** [comment]
Reasons for Using LightGBM and CatBoost
LightGBM:  
Efficiency: LightGBM is designed to be highly efficient and can handle large datasets with faster training times.
Accuracy: It often provides better accuracy compared to other gradient boosting algorithms.
Scalability: LightGBM can handle large-scale data and high-dimensional features.
Support for Categorical Features: It can directly handle categorical features without the need for one-hot encoding.
CatBoost:  
Handling Categorical Features: CatBoost is specifically designed to handle categorical features effectively, reducing the need for extensive preprocessing.
Robustness: It is less prone to overfitting and provides robust performance on various datasets.
Ease of Use: CatBoost requires minimal parameter tuning and is easy to use.
Efficiency: It is optimized for fast training and prediction times.
These models are particularly useful for tasks involving structured data and can provide significant improvements in performance and efficiency.


# Import necessary libraries
import lightgbm as lgb
from catboost import CatBoostRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, mean_absolute_percentage_error, accuracy_score, precision_score, confusion_matrix, recall_score, f1_score

# Function to train and evaluate a model
def train_and_evaluate_model(model, X_train, X_test, y_train, y_test):
    model.fit(X_train, y_train)
    pred = model.predict(X_test)
    rmse = np.sqrt(mean_squared_error(y_test, pred))
    mae = mean_absolute_error(y_test, pred)
    mape = mean_absolute_percentage_error(y_test, pred)
    accuracy = accuracy_score(y_test > pred, y_test > pred.round())
    precision = precision_score(y_test > pred, y_test > pred.round())
    confusion = confusion_matrix(y_test > pred, y_test > pred.round())
    recall = recall_score(y_test > pred, y_test > pred.round())
    f1 = f1_score(y_test > pred, y_test > pred.round())
    return rmse, mae, mape, accuracy, precision, confusion, recall, f1

# Train and evaluate LightGBM model
model_lightgbm = lgb.LGBMRegressor()
metrics_lightgbm = train_and_evaluate_model(model_lightgbm, X_train, X_test, y_train, y_test)
print("LightGBM Metrics:", metrics_lightgbm)

# Train and evaluate CatBoost model
model_catboost = CatBoostRegressor(verbose=0)
metrics_catboost = train_and_evaluate_model(model_catboost, X_train, X_test, y_train, y_test)
print("CatBoost Metrics:", metrics_catboost)

**Comment by praveenarjun** [comment]
I am fasing some issue to merge and pull request

**Comment by praveenarjun** [comment]
To Improve Model fastness we can use LightGBM and CatBoost 
#15 
This is my Issue

**Comment by praveenarjun** [comment]
I used anpther branch for this so u can merge it fro there

**Comment by praveenarjun** [comment]
how much time it take to review the code and merge
